libdevice/device.h

```diff
@@ -17,9 +17,9 @@
 
 #if defined(__SPIR__) || defined(__SPIRV__) || defined(__NVPTX__)
 #ifdef __SYCL_DEVICE_ONLY__
-#define DEVICE_EXTERNAL SYCL_EXTERNAL __attribute__((weak))
+#define DEVICE_EXTERNAL SYCL_EXTERNAL
 #else // __SYCL_DEVICE_ONLY__
-#define DEVICE_EXTERNAL __attribute__((weak))
+#define DEVICE_EXTERNAL
 #endif // __SYCL_DEVICE_ONLY__
```

Comment:
This is required to get libdevice working, see [here](https://github.com/intel/llvm/pull/14096). I'm looking for a better solution for this, I just kept this here in case anybody plans on trying the prototype.

Reply:
I suppose importing devicelib symbols at compile step can be a solution (see https://github.com/intel/llvm/pull/15114).

On the other hand, I recall discussing the possibility of linking device libraries with upstream maintainers, who expressed a preference for shifting device library linking from the "compile" to the "link" step. It would be ideal if we could discover a solution that aligns with the long-term strategy of upstream and enables us to utilize the thinLTO framework for offload code linking, thereby avoiding the use of weak symbols.

Reply:
attention to @mdtoguchi who has been looking at importing devicelib at compile step from the SYCL perspective.
Point to note: During one of the LLVM community presentation, it was mentioned that they are trying to move importing devicelib to link time.

Reply:
As we already perform device library linking at link time we can consider abandoning the efforts to pull them into the compilation step.  My main concern with performing at the link step is the communication required from the driver to the `clang-linker-wrapper` informing which device libraries should be linked.  The less tie-in we have between the driver and the `clang-linker-wrapper` at link time, the better.  IMO, at the very least the linker wrapper should know a minimum default device libraries to link and any communication from the driver is manipulating that list.

Reply:
I agree with @mdtoguchi. Unless user wants to change the names/location or disable linking of device libraries, driver should not have any logic to handle device code linking other than invoking `clang-linker-wrapper`. It makes sense to have driver options for additional configuration of device libraries, but driver's implementation should be just passing corresponding values to `clang-linker-wrapper` where these options should be processed.

Reply:
That is very interesting.
While working on https://github.com/intel/llvm/pull/15114 I've been wondering whether there is a particular reason why we link against CUDA libdevice and libclc in the compile step, but also again in the link step.
Could I get some clarification on that?

Reply:
> That is very interesting. While working on #15114 I've been wondering whether there is a particular reason why we link against CUDA libdevice and libclc in the compile step, but also again in the link step. Could I get some clarification on that?

@Naghasan, @npmiller, are you able to help here?